libherokubuildpack/src/command.rs

```diff
@@ -0,0 +1,185 @@
+        &mut self,
+        stdout_write: OW,
+        stderr_write: EW,
+    ) -> io::Result<process::Output>;
+}
+
+impl CommandExt for process::Command {
+    fn spawn_and_write_streams<OW: Write + Send, EW: Write + Send>(
+        &mut self,
+        stdout_write: OW,
+        stderr_write: EW,
+    ) -> io::Result<process::Child> {
+        self.stdout(process::Stdio::piped())
+            .stderr(process::Stdio::piped())
+            .spawn()
+            .and_then(|child| write_child_process_output(child, stdout_write, stderr_write))
+    }
+
+    fn output_and_write_streams<OW: Write + Send, EW: Write + Send>(
+        &mut self,
+        stdout_write: OW,
+        stderr_write: EW,
+    ) -> io::Result<process::Output> {
+        let mut stdout_buffer = vec![];
+        let mut stderr_buffer = vec![];
+
+        self.spawn_and_write_streams(
+            tee(&mut stdout_buffer, stdout_write),
+            tee(&mut stderr_buffer, stderr_write),
+        )
+        .and_then(|mut child| child.wait())
+        .map(|status| process::Output {
+            status,
+            stdout: stdout_buffer,
+            stderr: stderr_buffer,
+        })
+    }
+}
+
+fn write_child_process_output<OW: Write + Send, EW: Write + Send>(
+    mut child: process::Child,
+    mut stdout_writer: OW,
+    mut stderr_writer: EW,
+) -> io::Result<process::Child> {
+    // Copying the data to the writers happens in separate threads for stdout and stderr to ensure
+    // they're processed in parallel. Example: imagine the caller uses io::stdout() and io::stderr()
+    // as the writers so that the user can follow along with the command's output. If we copy stdout
+    // first and then stderr afterwards, interleaved stdout and stderr messages will no longer be
+    // interleaved (stderr output is always printed after stdout has been closed).
+    //
+    // The rust compiler currently cannot figure out how long a thread will run (doesn't take the
+    // almost immediate join calls into account) and therefore requires that data used in a thread
+    // lives forever. To avoid requiring 'static lifetimes for the writers, we use crossbeam's
+    // scoped threads here. This enables writers that write, for example, to a mutable buffer.
+    unwind_panic(crossbeam::scope(|scope| {
+        let stdout_copy_thread = mem::take(&mut child.stdout)
+            .map(|mut stdout| scope.spawn(move |_| std::io::copy(&mut stdout, &mut stdout_writer)));
+
+        let stderr_copy_thread = mem::take(&mut child.stderr)
+            .map(|mut stderr| scope.spawn(move |_| std::io::copy(&mut stderr, &mut stderr_writer)));
+
+        let stdout_copy_result = stdout_copy_thread.map_or_else(|| Ok(0), join_and_unwind_panic);
+        let stderr_copy_result = stderr_copy_thread.map_or_else(|| Ok(0), join_and_unwind_panic);
+
+        // Return the first error from either Result or the child process value
+        stdout_copy_result.and(stderr_copy_result).map(|_| child)
+    }))
```

Comment:
I basically have no clue what's happening here. I'm not familiar with crossbeam. I'm aware it's a parallelism-ish library. 

To write this out it looks like we're using crossbeam scope which allows us to spawn scoped threads  https://docs.rs/crossbeam/0.7.3/crossbeam/fn.scope.html. We're moving the memory of stdou and stderr into those threads and essentially just copying whatever goes into stdout into whatever is passed in to the stdout_writer. I'm assuming this is blocking (inside of the thread).

Then we do some stuff which essentially boils down to joining the threads.

I'm not entirely following how this last line works, but it compiles so i'm sure it does:

```rust
        // Return the first error from either Result or the child process value
        stdout_copy_result.and(stderr_copy_result).map(|_| child)
```

Reply:
Notes for self: My naieve implementation is here https://github.com/heroku/buildpacks-ruby/blob/f33fd189414507645d563ceea675157c78847fc5/commons/src/env_command.rs#L330-L376. I used channels to get data out of the threads without lifetime issues.

Reply:
```
// Return the first error from either Result or the child process value
stdout_copy_result.and(stderr_copy_result).map(|_| child)
```

The [`and`](https://doc.rust-lang.org/std/result/enum.Result.html#method.and) does the following:
- If `stdout_copy_result` is an `Err`, return it and do nothing with `stderr_copy_result`.
- If `stdout_copy_result` is an `Ok`, return `stderr_copy_result` as is (so either an `Err` or an `Ok`).
- This means, after the `and`, we will have either the `Err` of `stdout_copy_result` or the `Ok` or `Err` of `stderr_copy_result`.

The `map` is a regular map, but in this case we actually don't mind what's in the `Ok`. If the `Result` after the `and` is `Ok` (meaning that both succeeded), we put `child` into the `Ok`, dropping whatever was in the `Ok` previously.

Reply:
> With crossbeam, we can convince the compiler to believe us that OW and EW are only referenced by the threads until the crossbeam scope ends.

That was my understanding as well. I think it's commented well enough. I'm just not familiar with the API so it was more of a talk-aloud. If my comments are generally correct there's nothing to update. Thanks for taking the time to add those annotations